contracts/shapella-upgrade/ShapellaUpgradeTemplate.sol

```diff
@@ -0,0 +1,528 @@
+        _verifySingleOZAdmin(_burner(), admin);
+    }
+
+    function _verifySingleOZAdmin(IAccessControlEnumerable accessControlled, address admin) internal view {
+        if (accessControlled.getRoleMemberCount(DEFAULT_ADMIN_ROLE) != 1
+         || accessControlled.getRoleMember(DEFAULT_ADMIN_ROLE, 0) != admin
+        ) {
+            revert WrongOZAccessControlAdmin(address(accessControlled));
+        }
+    }
+
+    function _prepareWithdrawalVault() internal {
+        _withdrawalVault().proxy__upgradeTo(_withdrawalVaultImplementation);
+        _withdrawalVault().initialize();
+    }
+
+    function _prepareAccountingOracle() internal {
+        (, uint256 epochsPerFrame, ) = _hashConsensusForAccountingOracle.getFrameConfig();
+        uint256 lastLidoOracleCompletedEpochId = _lidoOracle().getLastCompletedEpochId();
+
+        // NB: HashConsensus.updateInitialEpoch must be called after AccountingOracle implementation is bound to proxy
+        _accountingOracle().proxy__upgradeTo(_accountingOracleImplementation);
+        _hashConsensusForAccountingOracle.updateInitialEpoch(lastLidoOracleCompletedEpochId + epochsPerFrame);
+        _accountingOracle().initialize(
+            address(this),
+            address(_hashConsensusForAccountingOracle),
+            _accountingOracleConsensusVersion
+        );
+    }
+
+    function _prepareValidatorsExitBus() internal {
+        _validatorsExitBusOracle().proxy__upgradeTo(_validatorsExitBusOracleImplementation);
+        _validatorsExitBusOracle().initialize(
+            address(this),
+            address(_hashConsensusForValidatorsExitBusOracle),
+            _validatorsExitBusOracleConsensusVersion,
+            VEBO_LAST_PROCESSING_REF_SLOT
+        );
+    }
+
+    function _migrateLidoOracleCommitteeMembers() internal {
+        address[] memory members = _lidoOracle().getOracleMembers();
+        uint256 quorum = _lidoOracle().getQuorum();
+        bytes32 manage_members_role = _hashConsensusForAccountingOracle.MANAGE_MEMBERS_AND_QUORUM_ROLE();
+        bytes32 submit_data_role = _accountingOracle().SUBMIT_DATA_ROLE();
+
+        _hashConsensusForAccountingOracle.grantRole(manage_members_role, address(this));
+        _hashConsensusForValidatorsExitBusOracle.grantRole(manage_members_role, address(this));
+        for (uint256 i; i < members.length; ++i) {
+            _hashConsensusForAccountingOracle.addMember(members[i], quorum);
+            _accountingOracle().grantRole(submit_data_role, members[i]);
+
+            _hashConsensusForValidatorsExitBusOracle.addMember(members[i], quorum);
+            _validatorsExitBusOracle().grantRole(submit_data_role, members[i]);
+        }
+        _hashConsensusForAccountingOracle.renounceRole(manage_members_role, address(this));
+        _hashConsensusForValidatorsExitBusOracle.renounceRole(manage_members_role, address(this));
+    }
+
+    function _migrateDSMGuardians() internal {
```

Comment:
do we have any checks (maybe not onchain) that security parameters such as `maxDepositsPerBlock` or `pauseIntentValidityPeriodBlocks` are set correctly?